### ISSUE TITLE
Training of the lm head.

### ISSUE THREAD

**Issue by CuongNN218** [SRI-CSL/BayesAdapt#2]
Dear @colinski , 

Thanks for your contribution and such amazing work. 

I have a question related to the training of the LM head. As far as I understand, the weight of the LM head is optimized using the point estimation (similar to the A and B matrix of LORA), not stochastically, as the subspace parameters. However, in the line of code below, the lm head is forward N times to compute the logits? Is this necessary because I think if it is point-estimated, we only have to compute the features M times and then forward it to the LM head. If there is a distribution defined on the weight of the head, could you please point out where the code is defined?

I would truly appreciate it if you could clarify my confusion. I am looking forward to hearing from you.

https://github.com/SRI-CSL/BayesAdapt/blob/4a0b4d30e57a53f2f2679c4998800816596e266a/bayesadapt/trainers/trainer.py#L313

**Comment by colinski** [comment]
Actually in ScalaBL, (following pretty much all other works), we put a (stochastic) LoRA adapter around the ``lm_head``. The typical setup is to put stochastic LoRA's in the backbone and ``lm_head`` (see ``conf/lora/wrapper/default.yaml``). Then we set ``samples.test.backbone=10`` and ``samples.test.last_layer=1`` to get 10 total outputs.  Since the ``lm_head`` is stochastic, it will use a different sample from its posterior for each backbone sample.

 If you want to keep the ``lm_head`` a point estimate you can remove ``lm_head`` from the ``lora.wrapper.target_modules`` list in the config. It would still remain in the ``lora.config.target_modules`` list, allowing it to update via LoRA without being stochastic.

The main reason the code uses these nested sampling loops is to support last layer Bayesianization, where _only_ the ``lm_head`` is stochastic. In this case, we would remove the attention layers from the ``lora.wrapper.target_modules``, set ``samples.test.backbone=1``, and set ``samples.test.last_layer=10``.

Thanks for your interest in this work!

**Comment by CuongNN218** [comment]
@colinski Thank you for your response. I have one more question, as I see in the line 36 that the ELBO is a simple combination of the nll loss and the kl divergence without any scaling hyper-param beta as described in the paper. Is there any significant difference if we train the model with and without beta, or is this omitted elsewhere in the code?

https://github.com/SRI-CSL/BayesAdapt/blob/1952165e4076c12aab19ba981f45585fa05f8c41/bayesadapt/trainers/vi_trainer.py#L36

**Comment by colinski** [comment]
Yeah that is a subtle point that is not explained well in the paper. 
When training VI approaches like BLoB and ScalaBL, there is actually two optimizers and lr schedulers.
The first optimizer handles the standard NLL loss of the classifier, while the other only backprops the kl_loss from each LoRA layer.
The learning rate of this kl_optimizer is then equivalent to beta scaling on the kl_loss.
For this we use the lr scheduler introduced by BLoB (```BLoBKLScheduler``` in ```bayesadapt.schedulers.py```).
If you want more or less weight on the kl term in the loss, you can control it with the ```optim.kl_optimizer.lr``` field in the config.
And for what its worth, I've found that this lr scheduler is critical to getting good performance with BLoB and ScalaBL.
